jmbitcoin/jmbitcoin/secp256k1_transaction.py

```diff
@@ -388,57 +415,145 @@ def address_to_script(addr):
 # Output script to address representation
 
 def is_p2pkh_script(script):
-    if script[:3] == b'\x76\xa9\x14' and script[-2:] == b'\x88\xac' and len(
+    """ Given a script as bytes, returns True if the script
+    matches the pay-to-pubkey-hash pattern (using opcodes),
+    otherwise returns False.
+    """
+    if not isinstance(script, bytes):
+        script = binascii.unhexlify(script)
+    if script[:3] == P2PKH_PRE and script[-2:] == P2PKH_POST and len(
             script) == 25:
         return True
     return False
 
 def is_segwit_native_script(script):
-    """Is scriptPubkey of form P2WPKH or P2WSH"""
-    if script[:2] in [b'\x00\x14', b'\x00\x20']:
+    """Is script, as bytes, of form P2WPKH or P2WSH;
+    see BIP141 for definitions of 2 current valid scripts.
+    """
+    if not isinstance(script, bytes):
+        script = binascii.unhexlify(script)
+
+    if (script[:2] == P2WPKH_PRE and len(script) == 22) or (
+        script[:2] == P2WSH_PRE and len(script) == 34):
         return True
     return False
 
 def script_to_address(script, vbyte=0, witver=0):
+    """ Given a hex or bytes script, and optionally a version byte
+    (for P2SH) and/or a witness version (for native segwit witness
+    programs), convert to a valid address (either bech32 or Base58CE).
+    An important tacit assumption: anything which does not match
+    the parsing rules of native segwit, or pay-to-pubkey-hash, is
+    assumed to be p2sh.
+    Note also that this translates scriptPubKeys to addresses, not
+    redeemscripts to p2sh addresses; for that, use p2sh_scriptaddr.
+    """
     if not isinstance(script, bytes):
         script = binascii.unhexlify(script)
     if is_segwit_native_script(script):
-        #hrp interpreted from the vbyte entry, TODO this should be cleaner.
+        #hrp interpreted from the vbyte entry, TODO: better way?
         if vbyte in [0, 5]:
             hrp = 'bc'
+        elif vbyte == 100:
+            hrp = 'bcrt'
         else:
             hrp = 'tb'
         return bech32addr_encode(hrp=hrp, witver=witver,
-                                 witprog=[ord(x) for x in script[2:]])
+        witprog=struct.unpack('{}B'.format(len(script[2:])).encode(
```

Comment:
The update here is actually a bugfix; without it, Py3 makers fail to process bech32 destination addresses correctly.